File: examples/Clojure/mtserver.clj
(ns mtserver
  (:refer-clojure :exclude [send])
  (:require [zhelpers :as mq])
  (:import [org.zeromq ZMQ$Context ZMQQueue]))

;;
;; Multithreaded Hello World server
;; Isaiah Peng <issaria@gmail.com>
;;

(defrecord Worker [^ZMQ$Context ctx]
  Runnable
  (run [this]
    (let [receiver (mq/socket ctx mq/rep)]
      (mq/connect receiver "inproc://workers")
      (while true
        (let [string (mq/recv-str receiver)]
          (println (format "Received request: [%s]." string))
          (Thread/sleep 1)
          (mq/send receiver "World\u0000")))
      (.close receiver))))

(defn -main []
  (let [ctx (mq/context 1)
        clients (mq/socket ctx mq/router)
        workers (mq/socket ctx mq/dealer)]
    (mq/bind clients "tcp://*:5555")
    (mq/bind workers "inproc://workers")
    (dotimes [i 5]
      (-> ctx Worker. Thread. .start))
    (let [queue (ZMQQueue. ctx clients workers)]
      (.run queue))
    ;; We never get here..
    (.close clients)
    (.close workers)
    (.term ctx)))

File: examples/Rust/mtserver.rs
use std::{thread, time};

fn worker_routine(context: &zmq::Context) {
    let receiver = context.socket(zmq::REP).unwrap();
    assert!(receiver.connect("inproc://workers").is_ok());

    loop {
        let string = receiver.recv_string(0).unwrap().unwrap();
        println!("Received request: {}", string);
        thread::sleep(time::Duration::from_secs(1));
        receiver.send("World", 0).unwrap();
    }
}

fn main() {
    let context = zmq::Context::new();

    let clients = context.socket(zmq::ROUTER).unwrap();
    assert!(clients.bind("tcp://*:5555").is_ok());

    let workers = context.socket(zmq::DEALER).unwrap();
    assert!(workers.bind("inproc://workers").is_ok());

    for _ in 0..5 {
        let ncontext = context.clone();
        thread::spawn(move || worker_routine(&ncontext));
    }

    zmq::proxy(&clients, &workers).unwrap();
}
